### DISCUSSION TITLE
Typescript for useRouter

### DISCUSSION THREAD

**Discussion by IRediTOTO** [vercel/next.js#28622] [Help]
Hi, is there anyway implement type for `useRouter`?
I just want clearly type in code, I want to set up what should inside the `router.query`

```
is there a type like this ?
    const router = useRouter<{query:{productId:string}}>() 

```

**Comment by tatethurston** [comment]
I wrote [nextjs-routes](https://github.com/tatethurston/nextjs-routes) to enable stricter type checking for `next/link` and `next/router`. The [query example](https://github.com/tatethurston/nextjs-routes#query) achieves something similar to what you're asking for:

```ts
import { useRouter } from "next/link";

// query is now typed as `{ foo: string }`
const { query } = useRouter<"/foos/[foo]">();
```

**Reply by IRediTOTO** [reply]
This so cool. I will use it, but can you export it to an interface too? Sometimes need use like this

```
const myText=`this is a link ${ROUTES.ITEM({itemId:"abcdefg"})} `

```

**Reply by tatethurston** [reply]
Moving this thread to https://github.com/tatethurston/nextjs-routes/issues/38

**Reply by sachinraja** [reply]
This is the best typed routing solution I've used for Next.js. There's not too much difficulty in adding it, it's pretty much just drop-in.
